ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_shared_utils/platform.py

```diff
@@ -162,23 +162,46 @@ def default_interfaces(self: SharedUtils) -> dict:
         default_interfaces set based on default_interfaces
         """
         default_interfaces = get(self.hostvars, "default_interfaces", default=[])
-
         device_platform = default(self.platform, "default")
 
+        # Re-order default_interfaces:
+        # #  platform  type
+        # 1  abc       xyz
+        # 2  abc       default
+        # 3  default   xyz
+        # 4  default   default
+        default_interfaces.sort(key=self.sort_default_interfaces)
+
         # First look for a matching default interface set that matches our platform and type
         for default_interface in default_interfaces:
             for platform in default_interface.get("platforms", []):
-                if search(f"^{platform}$", device_platform) and self.type in default_interface.get("types", []):
+                if search(f"^{platform}$", device_platform) and (
+                    self.type in default_interface.get("types", []) or "default" in default_interface.get("types", [])
+                ):
                     return default_interface
 
         # If not found, then look for a default default_interface that matches our type
         for default_interface in default_interfaces:
             for platform in default_interface.get("platforms", []):
-                if search(f"^{platform}$", "default") and self.type in default_interface.get("types", []):
+                if search(f"^{platform}$", "default") and (self.type in default_interface.get("types", []) or "default" in default_interface.get("types", [])):
                     return default_interface
 
         return {}
 
+    @staticmethod
+    def sort_default_interfaces(default_interface: dict) -> int:
+        """
+        Set sort order for each default_interface, based on platforms and types keys
+        """
+        if "default" not in default_interface.get("platforms") and "default" not in default_interface.get("types"):
+            return 1
+        elif "default" not in default_interface.get("platforms") and "default" in default_interface.get("types"):
+            return 2
+        elif "default" in default_interface.get("platforms") and "default" not in default_interface.get("types"):
+            return 3
+        else:
+            return 4
```

Comment:
so how do we end up sorting things like

```
  - types: [ spine ]
    platforms: [ "7050[SC]X3" ]
    uplink_interfaces: [ Ethernet49-52/1 ]
    mlag_interfaces: [ Ethernet53-54/1 ]
    downlink_interfaces: [ Ethernet1-30/1 ]

  - types: [ l3leaf, spine ]
    platforms: [ "7050[SC]X3" ]
    uplink_interfaces: [ Ethernet27-34 ]

  - types: [ spine, l3leaf ]
    platforms: [ "7050[SC]X3" ]
    uplink_interfaces: [ Ethernet42-48 ]
```

is it deterministic if I reverse the order of declaration ? or do we configure this config as wrong ? maybe we can sort the platform /type lists?

For this logic I am feeling like it could be worth writing some python tests with some inputs and expected outputs.

Reply:
Yup, I see what you're saying there.  That's definitely worth adding to make it a little more deterministic.

Reply:
Actually, scratch my earlier comment.  If we're just using "in" to verify if the node's type is 'in' the types list, why do we care about the order of those lists?  What matters is the order of the sorted default_interfaces list.  The sort routine currently used preserves the order in the original list if two identical sort keys are returned.  Is this deterministic? No.  If you change the order of the two conflicting items, you will get different results.  However I don't think we can save users from every single mistake.  The sorting applied was designed to ensure that anything with a 'default' in it, was matched last.  This satisfies the original use case of being able to define one default/default option, and then override it as needed.  I think beyond that we need to make it clear that once we find a platform that matches and then a type that matches, we stop looking.

Reply:
ok but so
```
  - types: [ default, spine ]
    platforms: [ "7050[SC]X3" ]
    uplink_interfaces: [ Ethernet49-52/1 ]
    mlag_interfaces: [ Ethernet53-54/1 ]
    downlink_interfaces: [ Ethernet1-30/1 ]

  - types: [ l3leaf, spine ]
    platforms: [ "7050[SC]X3" ]
    uplink_interfaces: [ Ethernet27-34 ]
```

you sort -> my spine get the second stuff only because it was with default

Reply:
and yes I understand this is how it works and I even find it quite logical.
I would ask if we are guaranteed that the order of keys is respected when sorting keys with the same value for all python versions we support also (though I surely hope so)

Reply:
OK.  So `platforms` was an existing key so I can't change the behaviour of that.  At present "default" as a platform has no intrinsic meeting - it's just a `platform_settings` entry named that.  If we remove consideration of "default" in the `platforms` key and just focus on sorting by `types` then we could:
* Sort by any `types` that are not-default, then default. 
* Enforce (becuse "default" as a type is a new concept) that "default" has to be the only item in the list
** i.e. `if "default" in types and len(types) > 1: raise AristaAvdError()`

This would eliminate the ambiguity in your example above and would also preserve the order of the list, with the exception of the new "default" entry.

@ClausHolbechArista - would appreciate your input in this too please.